### ISSUE TITLE
PHP 8 support

### ISSUE THREAD

**Issue by finwe** [mpdf/mpdf#1263]
At the moment, use `dev-php8-support` constraint in composer to use mpdf version with PHP8 support.

----

Currently, mPDF works OK woth 8.0-alpha2 as-is, but testing is an issue as older Mockery version used for mocking in tests is no longer compatible (Reflection issues).

Unless tests are fixed, support will not be official. There are a few possible approaches to address this:

- Remove PHP 5.6 support and upgrade underlying libraries (not sure if sufficient with PHP 7.0 support)
- Replace Mockery with another usable mocking framework (unlikely)
- Fork Mockery and quickfix issues with Reflection (management issues)
- Rewrite parts of mPDF (especially internal support classes dependencies on mPDF) to remove the need for mocking in tests; related #58 and #1262 (this would also solve other issues such as #1227)

**Comment by jakejackson1** [comment]
Mocker v1.3 supports PHP5.6 or higher. Mocker v1.4 requires PHP7.3 or higher. You could try upgrade to Mocker 1.3 and see if that resolves the issue with reflections in v8. Otherwise, perhaps it's time to start considering that PHP version bump?

**Comment by finwe** [comment]
Good call, thanks! Mockery v1.3 works!

Now just MpdfAbsoluteHeaderFooter test fails, because it uses the same failing reflection with default PHPUnit MockBuilder. I was not yet able to make it work with Mockery.

**Comment by jakejackson1** [comment]
@finwe looks to be because the `StreamWrapperChecker` is under the `Mpdf\File` namespace, and we've only overridden `stream_get_wrappers` on the `Mpdf\Image` namespace. 

Adding the following code to the bottom of `ImageProcessorTest.php` will resolve the issue:

```
namespace Mpdf\File;

function stream_get_wrappers()
{
	return \Mpdf\Image\stream_get_wrappers();
}
```

**Comment by finwe** [comment]
Currently waiting on FPDI, as a hard dependency, to support PHP 8.

**Comment by finwe** [comment]
This PR implements all CI and composer-related changes https://github.com/mpdf/mpdf/pull/1324

**Comment by finwe** [comment]
PHPUnit compatibility will also be a large issue. As PHP 8 is supported by PHPUnit 9 only and its lowest supported PHP version is 7.3. I do not see changing mPDF compatibility to a minimum of 7.3 plausible, at least yet.

**Comment by phpfui** [comment]
It is not absolutely necessary to run unit tests under PHP 8.  Really all that is needed is to make sure everything runs correctly under PHP 8 and composer support is added.  This library is widely used, so making it PHP 8.0 compatible is a high priority.

One possibility is to have separate unit test source for PHP 8. Not ideal, but would work.  At some point, support for PHP less than 7.1 should stop. 5.6 can not hold back the whole world because some people refuse to upgrade. Branch the 5.6 code and let it stagnate. I recently ported a huge project from 5.6 to 7.4, and we will be moving to 8 as soon as this library supports it. Lots of good tools to do so now.

I would be happy to help. Let me know of anything you would like done.

**Comment by finwe** [comment]
How do you propose _making sure everything runs correctly under PHP 8_ without running tests on PHP 8? :) Support checks **must** be continuous.

Yes, the library is widely used, but that is not a valid argument for PHP 8 support to be a high priority. Vast majority of its users will run a lesser version anyway, I'd wager.

General discussion about PHP versions support of the library is a bit off-topic here.

**Comment by phpfui** [comment]
I see two options for PHP 8.0 support.

1. Create a branch for older versions.  People can backport to that branch if they need / want future enhancements.  Main branch is then upgraded to more recent PHPUnit.  This also allows for the library to start using newer PHP enhancements, such as typed scalar parameters and return values.  This is a major enhancement to PHP and leads to better code.  What version of PHP to support on the main branch can be decided in another discussion, but it would be at least 7.0, but maybe 7.1, depending on the PHPUnit upgrade work.

2. Support two different versions of PHPUnit test code.  The current test code would test for the PHP version, then include the correct file (old tests or new tests).  I have not looked into the test code or much of the source code, but I imagine the source will run fine in PHP 8, as it looks pretty clean from what I have seen.  The disadvantage of this approach is maintaining two copies of the test code.

Personally I would go with option 1.  In either case, PHPUnit will have to be upgraded.  I am happy to do that work. I have found [https://github.com/rectorphp/rector](Rector) is pretty good at refactoring code.  I figure it would take an evening to do.

Let me know which way you want to go, and I can create a PR.  Option 1. would be the simplest, as the upgrade can be done inline, but happy to do option 2.  We would just have to figure out the directory structure.

**Comment by finwe** [comment]
With constraints given by https://phpunit.de/supported-versions.html, the minimal version should be 7.3 for PHPUnit to support PHP 8. That is a stretch way too far.

I see an option number 3, which is to transition to another testing library covering at least PHP 7.1 along with 8.0. I have not done the research if any usable one exists (nette/tester seemed a bit more forthcoming in PHP versions support).

Furthermore, I am haunted by _backporting_ of anything in any project the size and overall (poor) code quality of mPDF. If the code is to progress, backporting will not be a viable option in terms of amount of work needed by both contributors and maintainers.

In all, sunsetting older PHP version support calls for a new major version anyway. There is a milestone with things that I'd like to make in to version 9.

And last but not least: FPDI is still a blocker, unless the code is rewritten to not need FPDI by default (which at some point it probably should be).

**Comment by phpfui** [comment]
I would not recommend moving away from PHPUnit, as it is the most widely used testing library.  Last thing you need is to have nette decide PHPUnit is better than their code, and drop support.  Then you again stuck.

For PHPUnit backward compatibility, I support multiple packages that use different versions of PHPUnit depending on the version of PHP being tested.  It depends on the PHPUnit functionality you need.  If you don't use newer features, then older versions of PHPUnit work fine.  And since you would be upgrading from an old PHPUnit codebase, you would probably not be using the newer incompatible features anyway.

I would be happy to do a test to port PHPUnit to 7.0.  Not sure if that is possible, but worth a try.  Again, have not looked at the code yet.  I like to get buy in from the package maintainers before doing work.

**Comment by finwe** [comment]
Your point with the most widely used testing library is good and valid, but goes against the _width_ of supported PHP versions.

PHPUnit upgrade should be fairly easy, just replace testcase for the namespaced version and fix a few errors. I have tried this a while ago. What is the point though, if the goal is to support PHP 8?

**Comment by phpfui** [comment]
PHP 8 will not run on the older version of PHP unit. So to fully test PHP 8, you need to run a version of PHPUnit that will run on PHP 8.  The trick is to not use new PHPUnit features so the code runs on multiple versions of PHPUnit.

Composer will pick the most recent version of PHPUnit that will run on the version of PHP you are testing.  I do this in multiple projects.  Works great. I know for sure you can support 7.1 and higher including PHP 8.0, as I have several projects doing that now. Not sure about PHP 7.0, but worth a shot to try.  Don't think there is a version of PHPUnit that will work with the new namespaced PHPUnit code and still work on 5.6, but 5.6 does support namespaces, so not sure on that.  I generally only support PHP 7.1 or higher now, as the benefits of porting to PHP 7.1 far outweigh any possible reason for remaining on 5.6. Thankfully I stopped having to support PHP 5.6 (and 7.0) code 3 years ago.  My life improved measurably!

**Comment by finwe** [comment]
Yes, it is possible PHPUnit will run on officially unsupported version of PHP if you composer install it with another version of PHP. That's not a clean solution, though. If I am mistaken and this is somehow a pure, supported and maybe even documented option, I am all ears. If not, I'll repeat (just shuffle version numbers)

> How do you propose making sure everything runs correctly under PHP 8 without running tests on PHP 8? :) Support checks must be continuous.

Yes, my life would too improve measurably if I dropped support for PHP <7.3 but given mPDF userbase, that's not what I am willing to do. It would not improve lifes of mPDF's users.

**Comment by phpfui** [comment]
Option 2 should work then.  Just parallel versions of test code.  Not too hard to maintain both, as potentially only the namespace would change.  Maybe other minor things, but a good diff tool can easily move stuff over.  Would not require a major version change of the library.

But yes, long term, dropping PHP 5.6 would be a good thing.  And you don't have to really drop support, 5.6 users would just not get any more updates (unless really needed).

**Comment by marclaporte** [comment]
One option is to tag current version as LTS and just do security fixes (which are quite uncommon) on it and all new development / features happen on mPDF 9.x with a requirement of PHP 7.3+

Reasons: https://info.tiki.org/Versions

**Comment by phpfui** [comment]
@marclaporte +1!

**Comment by jakejackson1** [comment]
WordPress also has this problem and were able to to resolve it using phpunit 7.5 and some backported code. Relevant discussions here:

https://core.trac.wordpress.org/changeset/49037
https://core.trac.wordpress.org/ticket/50902
https://core.trac.wordpress.org/ticket/46149
https://core.trac.wordpress.org/changeset/48957

**Comment by phpfui** [comment]
@jakejackson1 Interesting, but looks like WordPress is using an image of PHPUnit and not using composer to resolve the dependency.  But good to know and thanks for the links.  And good to see WordPress is trying to get to PHP 7+.  WordPress is a major impediment to getting libraries updated to modern PHP versions.

**Comment by michaljusiega** [comment]
Probably PHPUnit 8.5 should have support for >= PHP7.2
See https://github.com/sebastianbergmann/phpunit/commit/73435f4bac40a63d18f04707858c738a64c1f7ed

**Comment by jakejackson1** [comment]
I've opened an issue in the FPDI repo about PHP8 support https://github.com/Setasign/FPDI/issues/123

**Comment by finwe** [comment]
FPDI is not as big a blocker as PHPUnit, though. It could be refactored out of the main mPDF class, I think, if necessary. 

And PHPUnit's just a political decision of the lowest supported PHP version. There is no good option.

**Comment by jakejackson1** [comment]
At a crossroads like this I think we should prioritise supporting new versions of PHP over supporting old versions of PHP. The idea is to move the project forward. People running the older versions of PHP will be happy to use an older version of mPDF until they are ready to upgrade PHP. But if we don't add support for it, there's no alternative for users running new versions of PHP.

**Comment by phpfui** [comment]
Having just ported another project to PHP 8.0, the lowest version of PHP that can be currently support along with PHP 8.0 with PHPUnit is PHP 7.1.  This is due to the fact that the oldest version of PHPUnit supported for PHP 8.0 is PHPUnit 8, and PHPUnit 8 uses return types (void) for setUp(), and PHP 7.0 does not support return types.

This really a diabolical plan by the author of PHPUnit to get everyone to stop supporting older versions of PHP.  It is actually a good thing, we need to get the PHP community to move forward.  Once you clear the PHP 7.0 hurdle (mostly mysql_ => mysqli_ stuff), then going to 8.0 is pretty easy.

I would go with @jakejackson1 and do a new version that supports 7.1 and higher. PHP 7.1 is a reasonable level of support.  7.2 and 7.3 don't have much to offer that is not in 7.1.  7.4 is a big improvement, but too far to go at this point.  Personally, I am going to love to just support 8.0 and higher at some point in the future, as there is so much goodness in 8.0.

**Comment by michaljusiega** [comment]
See https://github.com/Setasign/FPDI/compare/v2.3.4...v2.3.5 - fpdi has been released and allow install on PHP 8.0
CC @finwe

**Comment by iquito** [comment]
Why not make the new major version only support >=7.4 ? Even 7.3 is not actively supported anymore (only security fixes), and if the goal is to rewrite some of the code at some point the property types of 7.4 might come in handy. mpdf 8.x will remain available for anybody using 7.3 and below, and it seems unlikely that there will be much code added there, the only reason would be a security issue.

**Comment by finwe** [comment]
Because not supported does not mean not used.

**Comment by iquito** [comment]
Anybody using PHP 7.3 and below can use mpdf 8.x for as long as they want. If you are Wordpress, you want to support old PHP versions because you target end users who should upgrade often, even if they are using old PHP versions. With stable libraries like mpdf it is not important if users use old versions as long as there are no security issues. And mpdf is not changing rapidly, so if projects continue to use mpdf 8.x they will not be missing important features (unless I am misinterpreting the plans for mpdf).

What you gain by only supporting >=7.4:

- Less environments for tests, making tests easier to manage and faster to run
- Upgrade to PHPUnit 9, as even PHPUnit 8 is only supported until February 2021, PHPUnit 7 is already unsupported
- Less hassle in the future, when PHPUnit drops support for PHP 7.3, probably with PHPUnit 10
- Usage of new language features which can improve code quality

**Comment by phpfui** [comment]
@iquito is correct IMO. New version supports 7.4 and higher.  Branch the current version that will continue to support security fixes only for versions of PHP 7.3 and lower.  How may security fixes can there be?  Have there been a lot in the past?

Supporting 7.4 going forward allows for scalar parameter types and return values, plus member types, both of which are important features in moving the library forward.

One thing you want to avoid in any library is not updating things as they are released.  I recently took a look into upgrading an HTML Sanitizer.  The one I found looked really good and strict, but it did not support namespaces, as not doing autoloading well (used includes) and was completely dependent on SimpleTest, which has not been maintained for years.  In the end, it could not be saved because of too many obsolete dependencies and I ended up with another solution.  You don't want this to happen, trust me.

This is a good package with lots of users.  Users of older versions of PHP can use the older (pretty good version) with the occasional security fixes, and PHP 7.4 and higher users can use the more modern version with better type checking and improvements.

**Comment by finwe** [comment]
We are going around in circles.

Believe me, I am aware of all this, I have given this a decent bit of thought.

Of course the "correct" (and simplest) technocratic solution is to flush all that is old and unsupported. But is it really forthcoming to the user base? 

The quality of mPDF code is such that either you'd have to backport changes to older versions (maintenance hell) or to cut off older versions of PHP from fixes and (albeit minor) enhancements.

Providing mPDF userbase across a wider PHP version range with fixes and new features is, for me, more important than new language features, smaller builds, etc. etc.

If I am to freeze mPDF for older-PHP users, I'd like to do it in a state that is worthy of it. The higher the minimal PHP version would be, the better the state should be.

There are far, far nicer and more important things to be done than starting using types and many more than people on the bleeding edge versions can benefit from them.

**Comment by iquito** [comment]
If your goal is to support much of the user base of mpdf, you can also use the packagist statistics to decide: https://blog.packagist.com/php-versions-stats-2020-2-edition/

With PHP>=7.4, you would support ~42% of users
With PHP>=7.3, you would support ~70% of users
With PHP>=7.2, you would support ~85% of users

Keep in mind these numbers are changing fast, 7.4 went from 0% to 20% within the first 6 months (November 2019 until May 2020), and from 20% to 42% within the last 6 months, and PHP 8 will probably reach the same 20% usage within the next 6 months.

My intention was mainly to help the project itself, as many open source libraries get bogged down by backwards compatible support and then never have the resources to evolve. If I as a user had a vote, I would much prefer libraries focused on type safety going forward compared to many other things, as I have seen what a huge code quality improvement it has had in all my projects, especially together with static analyzers like Psalm and PHPStan. Using static analyzers for mpdf itself might also be something to consider, as it catches a lot of bugs early on.

**Comment by finwe** [comment]
Yes, With PHP >=7.2, I don't support 15 % of users, which in absolute numbers can be quite a lot. And that's not keeping in mind that the Packagist statistics are overall for the whole Packagist. I'd wager mPDF version summary would lean a bit more to the lesser versions. No, they don't have per-package statistics, I've asked.

Static analyzers are unable to cover mPDF at the moment (they mostly crash) given the state of its codebase. Believe me, I tried multiple times.

If you want to help the project itself, find an issue, try to solve it, send a pull request, help to move the project towards the state it can be frozen for old-PHP users without embarrassment.

**Comment by iquito** [comment]
It is your decision, of course, but it also makes it harder for people to contribute to mpdf if the dev dependencies are so old and very old PHP versions have to be supported, so changing the code base becomes even harder. I have supported many open source projects and have a few of my own, and having to support PHP < 7.2 would lead me to not contribute to a project just because of the additional headache. Backwards compatibility has a high price.

As a reference point why I actually joined this discussion: I wanted to upgrade one of my projects to PHP8, and mpdf is the only dependency that does not support it yet, which is how I stumbled upon this issue.

**Comment by finwe** [comment]
And again, I am aware of all of this.

Well, it's still better than to just push a PR with added `|| ~8.0.0` to PHP requirements, I'll give you that and thanks :)

**Comment by iquito** [comment]
Thanks for listening and for working on mpdf, I appreciate your effort! Looking forward to a PHP8 compatible release (I will be one of the early adopters) and maybe I can help out in some way in the future - I am already working on quite a bit of open source, so I am stretched thin currently, but the constructive discussion here is very encouraging. Thanks!

**Comment by phpfui** [comment]
I'll try to add a PHP 8.0 branch and PR this week.  Probably just a starting point, but I'll see what is involved.  I have ported some of it to my PHP 8.0 project, but if I found all the issues.

**Comment by finwe** [comment]
fyi, development branch (and the last release) runs on PHP 8 already, provided dependencies are installed with `--ignore-platform-reqs`. There is a `php8-support` branch with updated composer.json.

A few quirks still exist when running example repo on PHP8, but most of them in the examples source code itself.

**Comment by dakorpar** [comment]
> Yes, it is possible PHPUnit will run on officially unsupported version of PHP if you composer install it with another version of PHP. That's not a clean solution, though. If I am mistaken and this is somehow a pure, supported and maybe even documented option, I am all ears. If not, I'll repeat (just shuffle version numbers)
> 
> > How do you propose making sure everything runs correctly under PHP 8 without running tests on PHP 8? :) Support checks must be continuous.
> 
> Yes, my life would too improve measurably if I dropped support for PHP <7.3 but given mPDF userbase, that's not what I am willing to do. It would not improve lifes of mPDF's users.

I would say if someone is still on such old version pf php it most probably doesn't update dependencies as well since they're all moving forward so you're good however you turn it, they'll still have old version...

Supporting php 5.6 today is to much anyway and it's just slowing down progress of project... Minimum of minimum is at least 7.2...

**Comment by phpfui** [comment]
So had some time to spend on this.

On the php8-support branch, I was able to run all the test successfully by adding this to the phpunit section:

~~~
	convertErrorsToExceptions="false"
~~~

Then the depreciation notices are not exceptions.  I would put in a PR for this, but it is pretty minor, and figured you could do it yourself, and that would probably be easier all around.

I had also done a port for my project, and it was the same as the php8-support branch, except for the version number, so we ended up with the same solution (removing some leading default parameters as I recall).

So looks good to me.

**Comment by t0sk-fak3r** [comment]
Ok, so when will this be merged into master?

**Comment by finwe** [comment]
There is no ETA.

**Comment by dakorpar** [comment]
@finwe can we somehow help? IMHO this should have apsolute priority, there's some time already that php 8 is out and a lot of projects are stuck without possiblity to update to php 8 and use shiny new features...

**Comment by finwe** [comment]
> a lot of projects are stuck without possiblity to update to php 8 and use shiny new features...

Not true, `dev-php8-support` can be used. It is as good/bad a solution as any of the others.

**Comment by iquito** [comment]
@dakorpar Just use composer with `--ignore-platform-reqs`, this will ignore the "maximum" PHP version of packages. composer will probably also introduce a flag to specifically ignore upper limit PHP versions for packages, as those mostly don't make sense and have become a problem with PHP8. I am using mpdf with PHP8 without any problems so far.

**Comment by dakorpar** [comment]
I don't like idea having dev-* dependency in composer for production :confused: 
but sure it can work as temporary workaround

**Comment by finwe** [comment]
I believe using the branch constraint is better than `--ignore-platform-reqs` asi it will not obscure other potential compatibility errors.

**Comment by iquito** [comment]
@finwe: What other compatibility errors? It only checks the platform, meaning the PHP version and its extensions. If you are using PHP8 the minimum PHP version does not matter, only the maximum one like with mpdf in this case. And it can be replaced with the new flag from composer once that exists. The only thing you will be "missing" is that composer is not failing if a mandatory extension according to composer.json is missing, which seems better than including non-versioned dependencies.

**Comment by finwe** [comment]
Compatibility errors for other dependencies.

**Comment by iquito** [comment]
You could solve this by just relaxing the PHP version to >=5.6 and releasing a new version - this is what most projects did in the last months. Currently if you install mpdf on PHP8, it installs mpdf v6.1.3, which seems clearly worse.

**Comment by finwe** [comment]
> You could solve this by just relaxing the PHP version to >=5.6 and releasing a new version

This will lead exactly to the mpdf v6.1.3 situation in the future - installing a version that is not working with then-current PHP, so no. I know (think) Symfony went with this approach, but I think it is inherently wrong (and there is the `--ignore-platform-reqs` for testing these use cases).

The v6.1.3 situation is a sad legacy of not-knowing-better then. I also don't see a good way to mitigate this except changing all past tags allowing `>=5.6` which is a wrong thing to do by itself. But I hope this is a temporary situation until the PHP 8 support comes to a close (in my head mostly).

**Comment by iquito** [comment]
If that is really your opinion, you are still doing it wrong, as currently the restriction is `^5.6  || ^7.0`. Clearly that restriction has been there before PHP 7.4 or 7.3 were released, yet there you didn't think the new PHP version was a problem. PHP8 might be a major version number, but the number of changes in minor versions (including BC breaks) are just as numerous - just look at https://www.php.net/manual/en/migration74.incompatible.php or https://www.php.net/manual/en/migration73.incompatible.php . I had more trouble upgrading within 7.x than now to 8.0 because of some of the BC breaks in 7.1, 7.2, 7.3 and 7.4.

You are preventing people from trying your library with a new version, and if there are problems, you want people to report them earlier rather than later. That you will find all issues with PHP8 before releasing a "PHP8 compatible version" is highly unlikely, or just as likely as finding all bugs in software in general. You are mainly slowing down finding those bugs, and telling people to use a non-versioned tag of mpdf (`dev-php8-support`) has been viewed as bad software practice for a long time.

**Comment by finwe** [comment]
No. https://github.com/mpdf/mpdf/commit/f4ea96284f61b346f601194b7547a900bbf7b2c0

**Comment by finwe** [comment]
> ou are preventing people from trying your library with a new version
Again, ignore platform reqs.

**Comment by iquito** [comment]
> No. [f4ea962](https://github.com/mpdf/mpdf/commit/f4ea96284f61b346f601194b7547a900bbf7b2c0)

I guess I was wrong about that.

> Again, ignore platform reqs.

Most people won't or can't do this, just like they won't use a git tag like `dev-master` or `dev-php8-support`. And all these solutions have drawbacks / side effects.

**Comment by finwe** [comment]
Oh come on.

The situation is simple: unless I can guarantee the PHP version works well including CI checks, I will not mark it as supported.

Yes, there are workarounds to make it work, choose one that is best for you but know its drawbacks.

Anything else is moot.

**Comment by iquito** [comment]
Anybody using a new PHP version within the first 2-3 months knows they should expect things to go wrong - both with PHP itself and with existing PHP code. From my perspective it seems better to notice library bugs early and be able to report/fix them (and even know about them) than to just say it is not supported yet and make it harder for people to try it out, but I guess we do not agree in that.

**Comment by finwe** [comment]
Yes, let's agree to disagree.

**Comment by finwe** [comment]
8.0.10 with PHP 8 support released.

Let's continue the shitstorm with 8.1, that will be quite more interesting.

**Comment by jakejackson1** [comment]
@finwe this should help with running the automated tests from PHP5.6 to PHP8.0: https://packagist.org/packages/yoast/phpunit-polyfills

Will require a bit of grunt work converting the existing tests though.

**Comment by ToBulk** [comment]
I'm trying to get compatibility with PHP 8.x with my WordPress plugin DKPDF. This uses mPDF.

This no longer works with PHP 8.x..

I can not see what it was changed in the version discussed here so that the compatibility is made.

Thanks!

**Comment by ToBulk** [comment]
I'm trying to get compatibility with PHP 8.x with my WordPress plugin DKPDF. This uses mPDF.

[https://de.wordpress.org/plugins/dk-pdf/](url)

This no longer works with PHP 8.x..

I can not see what it was changed in the version discussed here so that the compatibility is made.

I am happy to receive help. Thanks!

**Comment by finwe** [comment]
@ToBulk this is not a problem of mPDF, refer to creators of DKPDF, they do not support PHP 8: https://github.com/Dinamiko/dk-pdf
